### ISSUE TITLE
Vim async feature for parallel update

### ISSUE THREAD

**Issue by chrisbra** [junegunn/vim-plug#498]
Does vim-plug use the async feature of recent Vim builds for parallel fetching of plugins?
If so, this should be documented somewhere, if not, could this be please be enabled?

```
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled May 10 2016 21:39:57)
Included patches: 1-1829
Compiled by cb@256bit.org
Huge version with GTK2 GUI.  Features included (+) or not (-):
+acl             +farsi           +mouse_netterm   +tag_binary
+arabic          +file_in_path    +mouse_sgr       +tag_old_static
+autocmd         +find_in_path    -mouse_sysmouse  -tag_any_white
+balloon_eval    +float           +mouse_urxvt     -tcl
+browse          +folding         +mouse_xterm     +termguicolors
++builtin_terms  -footer          +multi_byte      +terminfo
+byte_offset     +fork()          +multi_lang      +termresponse
+channel         +gettext         -mzscheme        +textobjects
+cindent         -hangul_input    +netbeans_intg   +timers
+clientserver    +iconv           +packages        +title
+clipboard       +insert_expand   +path_extra      +toolbar
+cmdline_compl   +job             -perl            +user_commands
+cmdline_hist    +jumplist        +persistent_undo +vertsplit
+cmdline_info    +keymap          +postscript      +virtualedit
+comments        +langmap         +printer         +visual
+conceal         +libcall         +profile         +visualextra
+cryptv          +linebreak       +python/dyn      +viminfo
```

- Type:
    - [ ] Bug
    - [ ] Enhancement
    - [ ] Feature Request
    - [x] Question
- OS:
    - [x] All/Other
    - [ ] Linux
    - [ ] OS X
    - [ ] Windows
- Vim:
    - [x] Terminal Vim
    - [x] GVim
    - [ ] Neovim

**Comment by junegunn** [comment]
It's not on my priority list as we already have Ruby and Python installer for Vim. Although they are not non-blocking, I don't really see the benefit of a non-blocking installer as the process usually finishes pretty quickly (`:PlugUpdate99`) and the asynchronous nature can be a problem in some contexts (see #104). Plus, it's hard to find a Vim installation without `+python` or `+ruby` these days.

The problem I see is that vim-plug currently have three versions of parallel installer - Ruby, Python, and the one using Neovim job API - and adding yet another version will cause extra maintenance burden. Hopefully in the end, Neovim team will embrace the new job API of Vim, and then we can update the current Neovim installer to use it so that it works both on Vim and Neovim.

**Comment by chrisbra** [comment]
okay fair enough. BTW: While I often see on Windows a Vim with `+python/dyn` or `+ruby/dyn` one is usually missing the runtime dlls so that they usually don't work as those interpreters are usually not available there.

Nevertheless, you might want to document at least, that vims async feature isn't supported yet.

**Comment by apriendeau** [comment]
@junegunn @chrisbra I completely agree that using 3 versions is a lot. It is another thing to ignore the core change in VIM itself. I would think it would be beneficial to swap out one of the 3 for core vim change (i.e. drop ruby or python). Moving to support only native features now that they are available? 

What about some kind of plugin architecture for this so people can choose? I would think that also defaulting the vim or Neovim's async should be first then using ruby and/or python as a backup would be the logical steps forward. 

These might more long term thoughts that immediate but none the less, something to chew on. 
¯\_(ツ)_/¯

**Comment by nfischer** [comment]
> I would think that also defaulting the vim or Neovim's async should be first then using ruby and/or python as a backup would be the logical steps forward.

@apriendeau I believe that neovim's async feature actually is chosen first for nvim users.

I agree that vim's async probably should be supported in the future, but some systems will always be stuck on older versions without vim's async.

**Comment by apriendeau** [comment]
@nfischer I agree. I would assume it would be the fallback system. The order of operations being nvim -> vim -> third-party clients so that we can use neovim's first.

**Comment by agauniyal** [comment]
> process usually finishes pretty quickly

Well not for everyone...

**Comment by apriendeau** [comment]
@agauniyal what issues you seeing?

**Comment by agauniyal** [comment]
No issues with vim or vim-plug with itself, I was just pointing out that update process doesn't finishes quickly for everyone. Some of us have limited internet bandwidth/speed and some plugins do take a lot of time to finish building/compiling. That's the only reason I use nvim in place of vim but ycm doesn't works well with it, so I'm back with vim.

**Comment by apriendeau** [comment]
@agauniyal I use deoplete for neovim and it works amazing. That is very true point. Limited internet is a very valid issue.

**Comment by pdavydov108** [comment]
It would be very cool if vim-plug could run plugin install scripts asynchronously and/or in parallel. For example YouCompleteMe installation script with c++ and rust (racer) native completers enabled can run 5-10 minutes on my machine. Such massive rebuilds doesn't happen on every update of course. Other plugins such as color_coded or fzf also take some time to install or download components.

**Comment by junegunn** [comment]
@pdavydov108 

> install scripts asynchronously and/or in parallel

The install script of YCM takes much longer than those of the others, so we will save only a fraction of the total time even if we run scripts in parallel. But I guess it makes sense to run it asynchronously. It's not ideal but for now we can achieve that using Neovim terminal or dispatch.vim.

```vim
" Neovim (see https://github.com/junegunn/vim-plug/issues/521)
Plug 'Valloric/YouCompleteMe', 
\ { 'do': ':new \| call termopen(''./install.py --system-libclang --clang-completer'')' }

" Using dispatch.vim (e.g. tmux)
Plug 'tpope/vim-dispatch'
Plug 'Valloric/YouCompleteMe', { 'do': ':Dispatch ./install.py --system-libclang --clang-complete' }
```

**Comment by agauniyal** [comment]
@pdavydov108 You can use system clang and system boost options while building YCM so that it builds much faster. Goes down from 3 mins to 35-40secs for me. Just make sure both are latest version for your distro.

**Comment by vhda** [comment]
While I'm in favor of using the newer Vim features, I think it might not be very valuable in this specific scenario. I see two possible implementations:

1. YCM compilation is run parallel but vim-plug waits for it to complete before finishing - in this case we don't gain much when compared with the current implementation.
2. YCM compilation is run in background and vim-plug will finish even if the installation is not complete - we will end up running a Vim session with a plugin that has an incomplete installation.

Or am I missing something? :)

**Comment by agauniyal** [comment]
I cannot say for other plugins but YCM launches a new process for each instance, so no you won't end up running a vim session with a plugin which has incomplete installation unless you choose to close current installation of vim-plug.

**Comment by agauniyal** [comment]
and ycm takes much time to fetch a number of submodules as well as latest clang executable and boost sources I guess(not sure though).

**Comment by vheon** [comment]
> You can use system clang and system boost options while building YCM so that it builds much faster

I remember you that what you're describing is an unsupported configuration. We would probably throw away the majority of boost when we switch to C++11.

> and ycm takes much time to fetch a number of submodules

the major submodule that we update is [Valloric/ycmd](/Valloric/ycmd) and we update the sub-submodule usually only when necessary.

> as well as latest clang executable

We cache the libclang binaries so we download it the first time you need it but if you reinstall we use the cached version

> and boost sources I guess(not sure though).

The boost sources are embedded in [Valloric/ycmd](/Valloric/ycmd)

**Comment by junegunn** [comment]
I and @mattn are working on this on vim8 branch. 

All the tests are passing.

- https://travis-ci.org/junegunn/vim-plug/builds/168217454

If you're interested, try it in advance and let me know how it works for you.

- https://raw.githubusercontent.com/junegunn/vim-plug/vim8/plug.vim

**Comment by vhda** [comment]
As far as I can tell the update is done in 3 phases:

1. Plugin detection and queue for update.
2. Plugin update.
3. After update checks.

For me, it seems that only step 2 is working in parallel when running in Vim 8.

**Comment by junegunn** [comment]
@vhda That's right. The rational behind the decision can be found in #430. I don't think we need to parallelize phase 3 (probably just for YCM), but we can consider making it asynchronous as the next step once this is merged.

**Comment by vhda** [comment]
@junegunn from my single test run, I would say that it is more important to parallelize phase 1, to try to start `git pull` from phase 2 as soon as possible. Phase 3 didn't appear to influence total runtime that much.

**Comment by chrisbra** [comment]
Am 2016-10-17 11:17, schrieb Junegunn Choi:
> I and @mattn [1] are working on this on vim8 branch.

I appreciate it. Thanks.

**Comment by junegunn** [comment]
@vhda Is it noticeably slower than Ruby/Python installer? In my case, updating 59 plugins with async vim8 installer takes almost the same amount of time as Ruby version (12 ~ 13 sec). The difference becomes only noticeable when I greatly increase the concurrency (`:PlugUpdate99`).

**Comment by vhda** [comment]
@junegunn I think this might be an issue with the machine I am running vim on, so please disregard my last comment. Updating 40 plugins took:

* master branch: ~20s in user space (11s total)
* vim8 branch; first run: ~20s in user space (39s total)
* vim8 branch: ~20s in user space (11s total)

When I change to the vim8 branch it always takes extra time, but the total CPU time is the same. I have no idea why this happens. The remaining runs take as much (total) time as the master branch. Again, this might be related with this specific machine/OS/etc.

Note: in both cases, frequently multiple vim subprocesses are left hanging and I have to kill them manually to have vim-plug complete the update. This is something that always happened randomly, but it was quite more noticeable now that I ran the update process multiple times. Did anyone ever report something like this?

Vim version: 8.0 (1-12)
Command used to test: time vim -c 'PlugUpdate | qall!'

**Comment by mattn** [comment]
You need Vim 8.0.0039 or later because older vim8 doesn't handle the exited-job status immediately.

**Comment by junegunn** [comment]
@mattn Thanks for the info, looks like we should require that patch version, so 1bbb7a5.

Another issue I noticed is that `out_cb` is sometimes called after `exit_cb`. I don't know if it's normal or expected, but anyway I pushed another commit to handle such cases: d1022e948e19b0cb8f4db7ae48b736044074225f.

**Comment by chrisbra** [comment]
Using commit 1bbb7a55e349f8634078b20aa1875fe5f52ea56e i don't see much difference with a newly compiled vim [8.0.41}(https://gist.github.com/chrisbra/da08d7d00c92a35b6d97fcfd6247b35d)
Using the master branch: `time vim -c 'PlugUpdate' -c 'qall'` took 3 seconds, however usign the master branch takes also 3 seconds (for 43 plugins, someof which are no real vim plugins but rather large repositories)

**Comment by mattn** [comment]
BTW, as you know, system() blocks anything. How about to implement job-based system() like below?

```diff
function! s:system(cmd)
  let out = ''
  let job = job_start([&shell, &shellcmdflag, a:cmd], {'out_cb': {ch,msg->[execute("let out .= msg"), out]}, 'out_mode': 'raw'})
  while job_status(job) == 'run'
    sleep 1m
  endwhile
  return out
endfunction
```

While doing the loop above, it works parallel for another jobs. of course, we've better to implement queue-based system.

**Comment by junegunn** [comment]
@mattn Yeah, we can revisit system() when we decide to make s:update_finish asynchronous/parallel. But I'm not sure that the benefit justifies the added complexity and the risk of leaving plugins in inconsistent states.

**Comment by junegunn** [comment]
Merged to master. Let me know if any issue arises. Thanks.
